test/params_test.c

```diff
@@ -0,0 +1,222 @@
+        swap_copy(le, raw_values[n].value, len); \
+        memcpy(&in, buf, sizeof(in)); \
+        param.buffer = &out; \
+        if (!TEST_true(OSSL_PARAM_set_##name(&param, "a", in))) \
+            return 0; \
+        copy_to_le(cmp, &out, sizeof(out)); \
+        if (!TEST_mem_eq(cmp, len, le, len)) \
+            return 0; \
+        in = 0; \
+        param.buffer = buf; \
+        if (!TEST_true(OSSL_PARAM_get_##name(&param, "a", &in))) \
+            return 0; \
+        copy_to_le(cmp, &in, sizeof(in)); \
+        if (!TEST_mem_eq(cmp, sizeof(in), le, sizeof(in))) \
+            return 0; \
+        param.buffer = &out; \
+        return test_param_type_extra(&param, le, sizeof(type)); \
+    }
+
+/*
+ * The negative test needs to come from outside the macro to avoid
+ * unsigned comparisons that are always true.
+ */
+TEST_INT(int, int)
+TEST_INT(long, long int)
+TEST_INT(uint, unsigned int)
+TEST_INT(ulong, unsigned long int)
+TEST_INT(int32, int32_t)
+TEST_INT(uint32, uint32_t)
+TEST_INT(int64, int64_t)
+TEST_INT(uint64, uint64_t)
+TEST_INT(size_t, size_t)
+
+static int test_param_bignum(int n)
+{
+    unsigned char buf[MAX_LEN], bnbuf[MAX_LEN], le[MAX_LEN];
+    const size_t len = raw_values[n].len;
+    size_t bnsize;
+    BIGNUM *b = NULL, *c = NULL;
+    OSSL_PARAM param = OSSL_PARAM_DEFN("bn", OSSL_PARAM_UNSIGNED_INTEGER,
+                                       NULL, sizeof(bnbuf), NULL);
+    int ret = 0;
+
+    param.buffer = bnbuf;
+    param.return_size = &bnsize;
+
+    copy_be_to_native(buf, raw_values[n].value, len);
+    swap_copy(le, raw_values[n].value, len);
+    if (!TEST_ptr(b = BN_bin2bn(raw_values[n].value, (int)len, NULL)))
+        goto err;
+
+    if (!TEST_true(OSSL_PARAM_set_BN(&param, "bn", b))
+        || !TEST_mem_eq(bnbuf, len, buf, len)
+        || !TEST_true(OSSL_PARAM_get_BN(&param, "bn", &c))
+        || !TEST_BN_eq(b, c))
+        goto err;
+
+    ret = 1;
+err:
+    BN_free(b);
+    BN_free(c);
+    return ret;
+}
+
+static int test_param_real(void)
+{
+    double p;
+    OSSL_PARAM param = OSSL_PARAM_double("r", NULL);
+
+    param.buffer = &p;
+    return TEST_true(OSSL_PARAM_set_double(&param, "r", 3.14159))
+           && TEST_double_eq(p, 3.14159);
+}
+
+int setup_tests(void)
+{
+    ADD_ALL_TESTS(test_param_int, OSSL_NELEM(raw_values));
+    ADD_ALL_TESTS(test_param_long, OSSL_NELEM(raw_values));
+    ADD_ALL_TESTS(test_param_uint, OSSL_NELEM(raw_values));
+    ADD_ALL_TESTS(test_param_ulong, OSSL_NELEM(raw_values));
+    ADD_ALL_TESTS(test_param_int32, OSSL_NELEM(raw_values));
+    ADD_ALL_TESTS(test_param_uint32, OSSL_NELEM(raw_values));
+    ADD_ALL_TESTS(test_param_size_t, OSSL_NELEM(raw_values));
+    ADD_ALL_TESTS(test_param_int64, OSSL_NELEM(raw_values));
+    ADD_ALL_TESTS(test_param_uint64, OSSL_NELEM(raw_values));
+    ADD_ALL_TESTS(test_param_bignum, OSSL_NELEM(raw_values));
+    ADD_TEST(test_param_real);
+    return 1;
+}
```

Comment:
From a typical usage point of view, this test looks like a very unusual use of `OSSL_PARAM`.  I understand that it's designed to test the itty gritty details of the API rather than the grander scheme, but in that case, I would call this file something like `params_api_test.c`, `params_internal_test.c`, or `params_unit_test.c`, to make space for a higher level `params_test.c` that may test the more typical uses from an application and provider point of view.  Yes, I have plans.

Reply:
No problem.  I'll rename it.

Reply:
I'll be interested in seeing the plans :)